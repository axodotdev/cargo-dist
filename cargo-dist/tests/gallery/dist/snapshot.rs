use super::*;

pub struct Snapshots {
    settings: insta::Settings,
    name: String,
    payload: String,
}

impl DistResult {
    // Run cargo-insta on everything we care to snapshot
    pub fn snapshot(&self) -> Result<Snapshots> {
        // We make a single uber-snapshot for both scripts to avoid the annoyances of having multiple snapshots
        // in one test (necessitating rerunning it multiple times or passing special flags to get all the changes)
        let mut snapshots = String::new();

        for app in &self.apps {
            append_snapshot_file(
                &mut snapshots,
                app.shell_installer_path
                    .as_deref()
                    .and_then(|p| p.file_name())
                    .unwrap_or_default(),
                app.shell_installer_path.as_deref(),
            )?;
            append_snapshot_file(
                &mut snapshots,
                app.homebrew_installer_path
                    .as_deref()
                    .and_then(|p| p.file_name())
                    .unwrap_or_default(),
                app.homebrew_installer_path.as_deref(),
            )?;
            append_snapshot_file(
                &mut snapshots,
                app.powershell_installer_path
                    .as_deref()
                    .and_then(|p| p.file_name())
                    .unwrap_or_default(),
                app.powershell_installer_path.as_deref(),
            )?;
            append_snapshot_tarball(
                &mut snapshots,
                app.npm_installer_package_path
                    .as_deref()
                    .and_then(|p| p.file_name())
                    .unwrap_or_default(),
                app.npm_installer_package_path.as_deref(),
            )?;
            append_snapshot_file(
                &mut snapshots,
                app.unified_checksum_path
                    .as_deref()
                    .and_then(|p| p.file_name())
                    .unwrap_or_default(),
                app.unified_checksum_path.as_deref(),
            )?;
        }

        Ok(Snapshots {
            settings: snapshot_settings_with_gallery_filter(),
            name: self.test_name.to_owned(),
            payload: snapshots,
        })
    }
}

impl PlanResult {
    // Run cargo-insta on everything we care to snapshot
    pub fn snapshot(&self) -> Result<Snapshots> {
        // We make a single uber-snapshot for both scripts to avoid the annoyances of having multiple snapshots
        // in one test (necessitating rerunning it multiple times or passing special flags to get all the changes)
        let mut snapshots = String::new();

        append_snapshot_string(&mut snapshots, "dist-manifest.json", &self.raw_json)?;

        Ok(Snapshots {
            settings: snapshot_settings_with_gallery_filter(),
            name: self.test_name.to_owned(),
            payload: snapshots,
        })
    }
}

impl GenerateResult {
    // Run cargo-insta on everything we care to snapshot
    pub fn snapshot(&self) -> Result<Snapshots> {
        // We make a single uber-snapshot for both scripts to avoid the annoyances of having multiple snapshots
        // in one test (necessitating rerunning it multiple times or passing special flags to get all the changes)
        let mut snapshots = String::new();

        eprintln!("{:?}", self.github_ci_path);
        append_snapshot_file(
            &mut snapshots,
            self.github_ci_path
                .as_deref()
                .and_then(|p| p.file_name())
                .unwrap_or_default(),
            self.github_ci_path.as_deref(),
        )?;

        append_snapshot_file(&mut snapshots, "main.wxs", self.wxs_path.as_deref())?;

        Ok(Snapshots {
            settings: snapshot_settings_with_gallery_filter(),
            name: self.test_name.to_owned(),
            payload: snapshots,
        })
    }
}

impl Snapshots {
    pub fn snap(self) {
        self.settings.bind(|| {
            insta::assert_snapshot!(self.name, self.payload);
        })
    }

    pub fn join(mut self, other: Self) -> Self {
        self.payload.push_str(&other.payload);
        self
    }
}

pub fn snapshot_settings() -> insta::Settings {
    let mut settings = insta::Settings::clone_current();
    let snapshot_dir = Utf8Path::new(ROOT_DIR).join("tests").join("snapshots");
    settings.set_snapshot_path(snapshot_dir);
    settings.set_prepend_module_to_snapshot(false);
    settings
}

pub fn snapshot_settings_with_version_filter() -> insta::Settings {
    let mut settings = snapshot_settings();
    settings.add_filter(
        r"\d+\.\d+\.\d+(\-prerelease\d*)?(\.\d+)?",
        "1.0.0-FAKEVERSION",
    );
    settings
}

/// Only filter parts that are specific to the toolchains being used to build the result
///
/// This is used for checking gallery entries
pub fn snapshot_settings_with_gallery_filter() -> insta::Settings {
    let mut settings = snapshot_settings();
    settings.add_filter(r#""dist_version": .*"#, r#""dist_version": "CENSORED","#);
    settings.add_filter(
        r#""cargo_version_line": .*"#,
        r#""cargo_version_line": "CENSORED""#,
    );
    settings.add_filter(
        r"cargo-dist/releases/download/v\d+\.\d+\.\d+(\-prerelease\d*)?(\.\d+)?/",
        "cargo-dist/releases/download/vSOME_VERSION/",
    );
    settings.add_filter(r#"sha256 ".*""#, r#"sha256 "CENSORED""#);
    settings.add_filter(r#""sha256": .*"#, r#""sha256": "CENSORED""#);
    settings.add_filter(r#""sha512": .*"#, r#""sha512": "CENSORED""#);
    settings.add_filter(r#""version":"[a-zA-Z\.0-9\-]*""#, r#""version":"CENSORED""#);
    settings.add_filter(
        r#""build_environment": \{\n\s+"macos": \{\n\s+"os_version": ".+"\n\s+}\n\s+}"#,
        r#""build_environment": "indeterminate""#,
    );
    settings.add_filter(
        r#""build_environment": \{\n\s+"linux": \{\n\s+"glibc_version": \{\n\s+"major": .+\n\s+"series": .+\n\s+\}\n\s+}\n\s+}"#,
        r#""build_environment": "indeterminate""#,
    );
    settings.add_filter(
        r#""build_environment": "windows""#,
        r#""build_environment": "indeterminate""#,
    );
    settings.add_filter(
        r"[0-9a-f]{64} .(?<filename>([a-zA-Z0-9-_]+)(\.tar\.gz|\.pkg))",
        "CENSORED (see https://github.com/axodotdev/cargo-dist/issues/1477)  $filename",
    );
    settings
}

/// Filter anything that will regularly change in the process of a release
///
/// This is used for checking `main` against itself.
#[allow(dead_code)]
pub fn snapshot_settings_with_dist_manifest_filter() -> insta::Settings {
    let mut settings = snapshot_settings_with_version_filter();
    settings.add_filter(
        r#""announcement_tag": .*"#,
        r#""announcement_tag": "CENSORED","#,
    );
    settings.add_filter(
        r#""announcement_title": .*"#,
        r#""announcement_title": "CENSORED""#,
    );
    settings.add_filter(
        r#""announcement_changelog": .*"#,
        r#""announcement_changelog": "CENSORED""#,
    );
    settings.add_filter(
        r#""announcement_github_body": .*"#,
        r#""announcement_github_body": "CENSORED""#,
    );
    settings.add_filter(
        r#""announcement_is_prerelease": .*"#,
        r#""announcement_is_prerelease": "CENSORED""#,
    );
    settings.add_filter(
        r#""cargo_version_line": .*"#,
        r#""cargo_version_line": "CENSORED""#,
    );
    settings.add_filter(r#""sha256": .*"#, r#""sha256": "CENSORED""#);
    settings.add_filter(r#""sha512": .*"#, r#""sha512": "CENSORED""#);
    settings.add_filter(
        r#""build_environment": \{\n\s+"macos": \{\n\s+"os_version": ".+"\n\s+}\n\s+}"#,
        r#""build_environment": "indeterminate""#,
    );
    settings.add_filter(
        r#""build_environment": \{\n\s+"linux": \{\n\s+"glibc_version": \{\n\s+"major": .+\n\s+"series": .+\n\s+\}\n\s+}\n\s+}"#,
        r#""build_environment": "indeterminate""#,
    );
    settings.add_filter(
        r#""build_environment": "windows""#,
        r#""build_environment": "indeterminate""#,
    );

    settings
}

fn append_snapshot_tarball(
    out: &mut String,
    name: &str,
    src_path: Option<&Utf8Path>,
) -> Result<()> {
    use std::io::Read;

    // Skip snapshotting this file if absent
    let Some(src_path) = src_path else {
        return Ok(());
    };

    // We shove everything in a BTreeMap to keep ordering stable
    let mut results = BTreeMap::new();

    let file = LocalAsset::load_bytes(src_path)?;
    let gz_decoder = flate2::read::GzDecoder::new(&file[..]);
    let mut tar_decoder = tar::Archive::new(gz_decoder);
    let entries = tar_decoder.entries().expect("couldn't read tar");
    for entry in entries {
        let mut entry = entry.expect("couldn't read tar entry");
        if entry.header().entry_type() == tar::EntryType::Regular {
            let path = entry
                .path()
                .expect("couldn't get tarred file's path")
                .to_string_lossy()
                .into_owned();
            let mut val = String::new();
            entry
                .read_to_string(&mut val)
                .expect("couldn't read tarred file to string");
            results.insert(path, val);
        }
    }

    for (path, val) in &results {
        append_snapshot_string(out, &format!("{name}/{path}"), val)?;
    }
    Ok(())
}

fn append_snapshot_file(out: &mut String, name: &str, src_path: Option<&Utf8Path>) -> Result<()> {
    // Skip snapshotting this file if absent
    let Some(src_path) = src_path else {
        return Ok(());
    };

    let src = axoasset::LocalAsset::load_string(src_path)?;
    append_snapshot_string(out, name, &src)
}

fn append_snapshot_string(out: &mut String, name: &str, val: &str) -> Result<()> {
    use std::fmt::Write;

    writeln!(out, "================ {name} ================").unwrap();
    writeln!(out, "{val}").unwrap();
    Ok(())
}
