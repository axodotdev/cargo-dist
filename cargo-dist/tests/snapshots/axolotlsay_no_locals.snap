---
source: cargo-dist/tests/gallery/dist/snapshot.rs
expression: self.payload
snapshot_kind: text
---
================ sha256.sum ================
CENSORED (see https://github.com/axodotdev/cargo-dist/issues/1477)  source.tar.gz


================ dist-manifest.json ================
{
  "dist_version": "CENSORED",
  "announcement_tag": "v0.2.2",
  "announcement_tag_is_implicit": true,
  "announcement_is_prerelease": false,
  "announcement_title": "Version 0.2.2",
  "announcement_changelog": "```text\n         +----------------------------------+\n         | now with arm64 linux binaries!!! |\n         +----------------------------------+\n        /\n≽(◕ ᴗ ◕)≼\n```",
  "announcement_github_body": "## Release Notes\n\n```text\n         +----------------------------------+\n         | now with arm64 linux binaries!!! |\n         +----------------------------------+\n        /\n≽(◕ ᴗ ◕)≼\n```\n\n## Download axolotlsay 0.2.2\n\n|  File  | Platform | Checksum |\n|--------|----------|----------|\n| [axolotlsay-aarch64-apple-darwin.tar.xz](https://github.com/axodotdev/axolotlsay/releases/download/v0.2.2/axolotlsay-aarch64-apple-darwin.tar.xz) | Apple Silicon macOS | [checksum](https://github.com/axodotdev/axolotlsay/releases/download/v0.2.2/axolotlsay-aarch64-apple-darwin.tar.xz.sha256) |\n| [axolotlsay-x86_64-apple-darwin.tar.xz](https://github.com/axodotdev/axolotlsay/releases/download/v0.2.2/axolotlsay-x86_64-apple-darwin.tar.xz) | Intel macOS | [checksum](https://github.com/axodotdev/axolotlsay/releases/download/v0.2.2/axolotlsay-x86_64-apple-darwin.tar.xz.sha256) |\n| [axolotlsay-x86_64-pc-windows-msvc.zip](https://github.com/axodotdev/axolotlsay/releases/download/v0.2.2/axolotlsay-x86_64-pc-windows-msvc.zip) | x64 Windows | [checksum](https://github.com/axodotdev/axolotlsay/releases/download/v0.2.2/axolotlsay-x86_64-pc-windows-msvc.zip.sha256) |\n| [axolotlsay-x86_64-unknown-linux-gnu.tar.xz](https://github.com/axodotdev/axolotlsay/releases/download/v0.2.2/axolotlsay-x86_64-unknown-linux-gnu.tar.xz) | x64 Linux | [checksum](https://github.com/axodotdev/axolotlsay/releases/download/v0.2.2/axolotlsay-x86_64-unknown-linux-gnu.tar.xz.sha256) |\n\n",
  "releases": [
    {
      "app_name": "axolotlsay",
      "app_version": "0.2.2",
      "env": {
        "install_dir_env_var": "AXOLOTLSAY_INSTALL_DIR",
        "unmanaged_dir_env_var": "AXOLOTLSAY_UNMANAGED_INSTALL",
        "disable_update_env_var": "AXOLOTLSAY_DISABLE_UPDATE",
        "no_modify_path_env_var": "AXOLOTLSAY_NO_MODIFY_PATH",
        "github_base_url_env_var": "AXOLOTLSAY_INSTALLER_GITHUB_BASE_URL",
        "ghe_base_url_env_var": "AXOLOTLSAY_INSTALLER_GHE_BASE_URL"
      },
      "display_name": "axolotlsay",
      "display": true,
      "artifacts": [
        "source.tar.gz",
        "source.tar.gz.sha256",
        "sha256.sum",
        "axolotlsay-aarch64-apple-darwin.tar.xz",
        "axolotlsay-aarch64-apple-darwin.tar.xz.sha256",
        "axolotlsay-x86_64-apple-darwin.tar.xz",
        "axolotlsay-x86_64-apple-darwin.tar.xz.sha256",
        "axolotlsay-x86_64-pc-windows-msvc.zip",
        "axolotlsay-x86_64-pc-windows-msvc.zip.sha256",
        "axolotlsay-x86_64-unknown-linux-gnu.tar.xz",
        "axolotlsay-x86_64-unknown-linux-gnu.tar.xz.sha256"
      ],
      "hosting": {
        "github": {
          "artifact_base_url": "https://github.com",
          "artifact_download_path": "/axodotdev/axolotlsay/releases/download/v0.2.2",
          "owner": "axodotdev",
          "repo": "axolotlsay"
        }
      }
    }
  ],
  "artifacts": {
    "axolotlsay-aarch64-apple-darwin.tar.xz": {
      "name": "axolotlsay-aarch64-apple-darwin.tar.xz",
      "kind": "executable-zip",
      "target_triples": [
        "aarch64-apple-darwin"
      ],
      "assets": [
        {
          "name": "CHANGELOG.md",
          "path": "CHANGELOG.md",
          "kind": "changelog"
        },
        {
          "name": "LICENSE-APACHE",
          "path": "LICENSE-APACHE",
          "kind": "license"
        },
        {
          "name": "LICENSE-MIT",
          "path": "LICENSE-MIT",
          "kind": "license"
        },
        {
          "name": "README.md",
          "path": "README.md",
          "kind": "readme"
        },
        {
          "id": "axolotlsay-aarch64-apple-darwin-exe-axolotlsay",
          "name": "axolotlsay",
          "path": "axolotlsay",
          "kind": "executable"
        }
      ],
      "checksum": "axolotlsay-aarch64-apple-darwin.tar.xz.sha256"
    },
    "axolotlsay-aarch64-apple-darwin.tar.xz.sha256": {
      "name": "axolotlsay-aarch64-apple-darwin.tar.xz.sha256",
      "kind": "checksum",
      "target_triples": [
        "aarch64-apple-darwin"
      ]
    },
    "axolotlsay-x86_64-apple-darwin.tar.xz": {
      "name": "axolotlsay-x86_64-apple-darwin.tar.xz",
      "kind": "executable-zip",
      "target_triples": [
        "x86_64-apple-darwin"
      ],
      "assets": [
        {
          "name": "CHANGELOG.md",
          "path": "CHANGELOG.md",
          "kind": "changelog"
        },
        {
          "name": "LICENSE-APACHE",
          "path": "LICENSE-APACHE",
          "kind": "license"
        },
        {
          "name": "LICENSE-MIT",
          "path": "LICENSE-MIT",
          "kind": "license"
        },
        {
          "name": "README.md",
          "path": "README.md",
          "kind": "readme"
        },
        {
          "id": "axolotlsay-x86_64-apple-darwin-exe-axolotlsay",
          "name": "axolotlsay",
          "path": "axolotlsay",
          "kind": "executable"
        }
      ],
      "checksum": "axolotlsay-x86_64-apple-darwin.tar.xz.sha256"
    },
    "axolotlsay-x86_64-apple-darwin.tar.xz.sha256": {
      "name": "axolotlsay-x86_64-apple-darwin.tar.xz.sha256",
      "kind": "checksum",
      "target_triples": [
        "x86_64-apple-darwin"
      ]
    },
    "axolotlsay-x86_64-pc-windows-msvc.zip": {
      "name": "axolotlsay-x86_64-pc-windows-msvc.zip",
      "kind": "executable-zip",
      "target_triples": [
        "x86_64-pc-windows-msvc"
      ],
      "assets": [
        {
          "name": "CHANGELOG.md",
          "path": "CHANGELOG.md",
          "kind": "changelog"
        },
        {
          "name": "LICENSE-APACHE",
          "path": "LICENSE-APACHE",
          "kind": "license"
        },
        {
          "name": "LICENSE-MIT",
          "path": "LICENSE-MIT",
          "kind": "license"
        },
        {
          "name": "README.md",
          "path": "README.md",
          "kind": "readme"
        },
        {
          "id": "axolotlsay-x86_64-pc-windows-msvc-exe-axolotlsay",
          "name": "axolotlsay",
          "path": "axolotlsay.exe",
          "kind": "executable"
        }
      ],
      "checksum": "axolotlsay-x86_64-pc-windows-msvc.zip.sha256"
    },
    "axolotlsay-x86_64-pc-windows-msvc.zip.sha256": {
      "name": "axolotlsay-x86_64-pc-windows-msvc.zip.sha256",
      "kind": "checksum",
      "target_triples": [
        "x86_64-pc-windows-msvc"
      ]
    },
    "axolotlsay-x86_64-unknown-linux-gnu.tar.xz": {
      "name": "axolotlsay-x86_64-unknown-linux-gnu.tar.xz",
      "kind": "executable-zip",
      "target_triples": [
        "x86_64-unknown-linux-gnu"
      ],
      "assets": [
        {
          "name": "CHANGELOG.md",
          "path": "CHANGELOG.md",
          "kind": "changelog"
        },
        {
          "name": "LICENSE-APACHE",
          "path": "LICENSE-APACHE",
          "kind": "license"
        },
        {
          "name": "LICENSE-MIT",
          "path": "LICENSE-MIT",
          "kind": "license"
        },
        {
          "name": "README.md",
          "path": "README.md",
          "kind": "readme"
        },
        {
          "id": "axolotlsay-x86_64-unknown-linux-gnu-exe-axolotlsay",
          "name": "axolotlsay",
          "path": "axolotlsay",
          "kind": "executable"
        }
      ],
      "checksum": "axolotlsay-x86_64-unknown-linux-gnu.tar.xz.sha256"
    },
    "axolotlsay-x86_64-unknown-linux-gnu.tar.xz.sha256": {
      "name": "axolotlsay-x86_64-unknown-linux-gnu.tar.xz.sha256",
      "kind": "checksum",
      "target_triples": [
        "x86_64-unknown-linux-gnu"
      ]
    },
    "sha256.sum": {
      "name": "sha256.sum",
      "kind": "unified-checksum"
    },
    "source.tar.gz": {
      "name": "source.tar.gz",
      "kind": "source-tarball",
      "checksum": "source.tar.gz.sha256"
    },
    "source.tar.gz.sha256": {
      "name": "source.tar.gz.sha256",
      "kind": "checksum"
    }
  },
  "systems": {
    "plan:all:": {
      "id": "plan:all:",
      "cargo_version_line": "CENSORED"
      "build_environment": "indeterminate"
    }
  },
  "publish_prereleases": false,
  "force_latest": false,
  "ci": {
    "github": {
      "artifacts_matrix": {
        "include": [
          {
            "targets": [
              "aarch64-apple-darwin"
            ],
            "runner": "macos-13",
            "install_dist": "curl --proto '=https' --tlsv1.2 -LsSf https://github.com/axodotdev/cargo-dist/releases/download/vSOME_VERSION/cargo-dist-installer.sh | sh",
            "dist_args": "--artifacts=local --target=aarch64-apple-darwin",
            "cache_provider": "github",
            "install_cargo_auditable": "curl --proto '=https' --tlsv1.2 -LsSf https://github.com/rust-secure-code/cargo-auditable/releases/latest/download/cargo-auditable-installer.sh | sh"
          },
          {
            "targets": [
              "x86_64-apple-darwin"
            ],
            "runner": "macos-13",
            "install_dist": "curl --proto '=https' --tlsv1.2 -LsSf https://github.com/axodotdev/cargo-dist/releases/download/vSOME_VERSION/cargo-dist-installer.sh | sh",
            "dist_args": "--artifacts=local --target=x86_64-apple-darwin",
            "cache_provider": "github",
            "install_cargo_auditable": "curl --proto '=https' --tlsv1.2 -LsSf https://github.com/rust-secure-code/cargo-auditable/releases/latest/download/cargo-auditable-installer.sh | sh"
          },
          {
            "targets": [
              "x86_64-pc-windows-msvc"
            ],
            "runner": "windows-2019",
            "install_dist": "powershell -c \"irm https://github.com/axodotdev/cargo-dist/releases/download/vSOME_VERSION/cargo-dist-installer.ps1 | iex\"",
            "dist_args": "--artifacts=local --target=x86_64-pc-windows-msvc",
            "cache_provider": "github",
            "install_cargo_auditable": "powershell -c \"irm https://github.com/rust-secure-code/cargo-auditable/releases/latest/download/cargo-auditable-installer.ps1 | iex\""
          },
          {
            "targets": [
              "x86_64-unknown-linux-gnu"
            ],
            "runner": "ubuntu-20.04",
            "install_dist": "curl --proto '=https' --tlsv1.2 -LsSf https://github.com/axodotdev/cargo-dist/releases/download/vSOME_VERSION/cargo-dist-installer.sh | sh",
            "dist_args": "--artifacts=local --target=x86_64-unknown-linux-gnu",
            "cache_provider": "github",
            "install_cargo_auditable": "curl --proto '=https' --tlsv1.2 -LsSf https://github.com/rust-secure-code/cargo-auditable/releases/latest/download/cargo-auditable-installer.sh | sh"
          }
        ]
      },
      "pr_run_mode": "plan"
    }
  },
  "linkage": [],
  "upload_files": []
}

================ release.yml ================
# This file was autogenerated by dist: https://opensource.axo.dev/cargo-dist/
#
# Copyright 2022-2024, axodotdev
# SPDX-License-Identifier: MIT or Apache-2.0
#
# CI that:
#
# * checks for a Git Tag that looks like a release
# * builds artifacts with dist (archives, installers, hashes)
# * uploads those artifacts to temporary workflow zip
# * on success, uploads the artifacts to a GitHub Release
#
# Note that the GitHub Release will be created with a generated
# title/body based on your changelogs.

name: Release
permissions:
  "contents": "write"

# This task will run whenever you push a git tag that looks like a version
# like "1.0.0", "v0.1.0-prerelease.1", "my-app/0.1.0", "releases/v1.0.0", etc.
# Various formats will be parsed into a VERSION and an optional PACKAGE_NAME, where
# PACKAGE_NAME must be the name of a Cargo package in your workspace, and VERSION
# must be a Cargo-style SemVer Version (must have at least major.minor.patch).
#
# If PACKAGE_NAME is specified, then the announcement will be for that
# package (erroring out if it doesn't have the given version or isn't dist-able).
#
# If PACKAGE_NAME isn't specified, then the announcement will be for all
# (dist-able) packages in the workspace with that version (this mode is
# intended for workspaces with only one dist-able package, or with all dist-able
# packages versioned/released in lockstep).
#
# If you push multiple tags at once, separate instances of this workflow will
# spin up, creating an independent announcement for each one. However, GitHub
# will hard limit this to 3 tags per commit, as it will assume more tags is a
# mistake.
#
# If there's a prerelease-style suffix to the version, then the release(s)
# will be marked as a prerelease.
on:
  pull_request:
  push:
    tags:
      - '**[0-9]+.[0-9]+.[0-9]+*'

jobs:
  # Run 'dist plan' (or host) to determine what tasks we need to do
  plan:
    runs-on: "ubuntu-20.04"
    outputs:
      val: ${{ steps.plan.outputs.manifest }}
      tag: ${{ !github.event.pull_request && github.ref_name || '' }}
      tag-flag: ${{ !github.event.pull_request && format('--tag={0}', github.ref_name) || '' }}
      publishing: ${{ !github.event.pull_request }}
    env:
      GH_TOKEN: ${{ secrets.GITHUB_TOKEN }}
    steps:
      - uses: actions/checkout@v4
        with:
          submodules: recursive
      - name: Install dist
        # we specify bash to get pipefail; it guards against the `curl` command
        # failing. otherwise `sh` won't catch that `curl` returned non-0
        shell: bash
        run: "curl --proto '=https' --tlsv1.2 -LsSf https://github.com/axodotdev/cargo-dist/releases/download/vSOME_VERSION/cargo-dist-installer.sh | sh"
      - name: Cache dist
        uses: actions/upload-artifact@v4
        with:
          name: cargo-dist-cache
          path: ~/.cargo/bin/dist
      # sure would be cool if github gave us proper conditionals...
      # so here's a doubly-nested ternary-via-truthiness to try to provide the best possible
      # functionality based on whether this is a pull_request, and whether it's from a fork.
      # (PRs run on the *source* but secrets are usually on the *target* -- that's *good*
      # but also really annoying to build CI around when it needs secrets to work right.)
      - id: plan
        run: |
          dist ${{ (!github.event.pull_request && format('host --steps=create --tag={0}', github.ref_name)) || 'plan' }} --output-format=json > plan-dist-manifest.json
          echo "dist ran successfully"
          cat plan-dist-manifest.json
          echo "manifest=$(jq -c "." plan-dist-manifest.json)" >> "$GITHUB_OUTPUT"
      - name: "Upload dist-manifest.json"
        uses: actions/upload-artifact@v4
        with:
          name: artifacts-plan-dist-manifest
          path: plan-dist-manifest.json

  # Build and package all the platform-agnostic(ish) things
  build-global-artifacts:
    needs:
      - plan
    runs-on: "ubuntu-20.04"
    env:
      GH_TOKEN: ${{ secrets.GITHUB_TOKEN }}
      BUILD_MANIFEST_NAME: target/distrib/global-dist-manifest.json
    steps:
      - uses: actions/checkout@v4
        with:
          submodules: recursive
      - name: Install cached dist
        uses: actions/download-artifact@v4
        with:
          name: cargo-dist-cache
          path: ~/.cargo/bin/
      - run: chmod +x ~/.cargo/bin/dist
      # Get all the local artifacts for the global tasks to use (for e.g. checksums)
      - name: Fetch local artifacts
        uses: actions/download-artifact@v4
        with:
          pattern: artifacts-*
          path: target/distrib/
          merge-multiple: true
      - id: cargo-dist
        shell: bash
        run: |
          dist build ${{ needs.plan.outputs.tag-flag }} --output-format=json "--artifacts=global" > dist-manifest.json
          echo "dist ran successfully"

          # Parse out what we just built and upload it to scratch storage
          echo "paths<<EOF" >> "$GITHUB_OUTPUT"
          jq --raw-output ".upload_files[]" dist-manifest.json >> "$GITHUB_OUTPUT"
          echo "EOF" >> "$GITHUB_OUTPUT"

          cp dist-manifest.json "$BUILD_MANIFEST_NAME"
      - name: "Upload artifacts"
        uses: actions/upload-artifact@v4
        with:
          name: artifacts-build-global
          path: |
            ${{ steps.cargo-dist.outputs.paths }}
            ${{ env.BUILD_MANIFEST_NAME }}
  # Determines if we should publish/announce
  host:
    needs:
      - plan
      - build-global-artifacts
    # Only run if we're "publishing", and only if local and global didn't fail (skipped is fine)
    if: ${{ always() && needs.plan.outputs.publishing == 'true' && (needs.build-global-artifacts.result == 'skipped' || needs.build-global-artifacts.result == 'success') }}
    env:
      GH_TOKEN: ${{ secrets.GITHUB_TOKEN }}
    runs-on: "ubuntu-20.04"
    outputs:
      val: ${{ steps.host.outputs.manifest }}
    steps:
      - uses: actions/checkout@v4
        with:
          submodules: recursive
      - name: Install cached dist
        uses: actions/download-artifact@v4
        with:
          name: cargo-dist-cache
          path: ~/.cargo/bin/
      - run: chmod +x ~/.cargo/bin/dist
      # Fetch artifacts from scratch-storage
      - name: Fetch artifacts
        uses: actions/download-artifact@v4
        with:
          pattern: artifacts-*
          path: target/distrib/
          merge-multiple: true
      - id: host
        shell: bash
        run: |
          dist host ${{ needs.plan.outputs.tag-flag }} --steps=upload --steps=release --output-format=json > dist-manifest.json
          echo "artifacts uploaded and released successfully"
          cat dist-manifest.json
          echo "manifest=$(jq -c "." dist-manifest.json)" >> "$GITHUB_OUTPUT"
      - name: "Upload dist-manifest.json"
        uses: actions/upload-artifact@v4
        with:
          # Overwrite the previous copy
          name: artifacts-dist-manifest
          path: dist-manifest.json
      # Create a GitHub Release while uploading all files to it
      - name: "Download GitHub Artifacts"
        uses: actions/download-artifact@v4
        with:
          pattern: artifacts-*
          path: artifacts
          merge-multiple: true
      - name: Cleanup
        run: |
          # Remove the granular manifests
          rm -f artifacts/*-dist-manifest.json
      - name: Create GitHub Release
        env:
          PRERELEASE_FLAG: "${{ fromJson(steps.host.outputs.manifest).announcement_is_prerelease && '--prerelease' || '' }}"
          ANNOUNCEMENT_TITLE: "${{ fromJson(steps.host.outputs.manifest).announcement_title }}"
          ANNOUNCEMENT_BODY: "${{ fromJson(steps.host.outputs.manifest).announcement_github_body }}"
          REPO: "custom-owner/cool-repo"
          RELEASE_COMMIT: "${{ fromJson(steps.host.outputs.manifest).ci.github.external_repo_commit }}"
          # We need to ensure `gh` uses the token that can auth
          # to the external repo instead of the runner token.
          GH_TOKEN: ${{ secrets.GH_RELEASES_TOKEN }}
        run: |
          # Write and read notes from a file to avoid quoting breaking things
          echo "$ANNOUNCEMENT_BODY" > $RUNNER_TEMP/notes.txt

          gh release create "${{ needs.plan.outputs.tag }}" --repo "$REPO" --target "$RELEASE_COMMIT" $PRERELEASE_FLAG --title "$ANNOUNCEMENT_TITLE" --notes-file "$RUNNER_TEMP/notes.txt" artifacts/*

  announce:
    needs:
      - plan
      - host
    # use "always() && ..." to allow us to wait for all publish jobs while
    # still allowing individual publish jobs to skip themselves (for prereleases).
    # "host" however must run to completion, no skipping allowed!
    if: ${{ always() && needs.host.result == 'success' }}
    runs-on: "ubuntu-20.04"
    env:
      GH_TOKEN: ${{ secrets.GITHUB_TOKEN }}
    steps:
      - uses: actions/checkout@v4
        with:
          submodules: recursive
